flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/schema/FlinkTableFunctionImpl.scala

```diff
@@ -27,52 +27,73 @@ import org.apache.flink.api.common.typeinfo.{AtomicType, TypeInformation}
 import org.apache.flink.api.common.typeutils.CompositeType
 import org.apache.flink.table.api.TableException
 import org.apache.flink.table.calcite.FlinkTypeFactory
+import org.apache.flink.table.functions.utils.UserDefinedFunctionUtils
+import org.apache.flink.table.functions.{TableFunction => FlinkUDTF}
 
 /**
   * This is heavily inspired by Calcite's [[org.apache.calcite.schema.impl.TableFunctionImpl]].
   * We need it in order to create a [[org.apache.flink.table.functions.utils.TableSqlFunction]].
   * The main difference is that we override the [[getRowType()]] and [[getElementType()]].
+  *
+  * @param tableFunction The Table Function instance
+  * @param implicitResultType The implicit result type.
+  * @param evalMethod The eval() method of the [[tableFunction]]
   */
-class FlinkTableFunctionImpl[T](
-    val typeInfo: TypeInformation[T],
-    val fieldIndexes: Array[Int],
-    val fieldNames: Array[String],
-    val evalMethod: Method)
+class FlinkTableFunctionImpl[T](val tableFunction: FlinkUDTF[_],
```

Comment:
Can we just separate this implementation into 2 classes, one for which we already know the type, and one for whose type has not been decided yet. 
Why i suggesting this is that i noticed `tableFunction.getResultType` is called multiple times when evaluating the plan, even we had already determine the row type long time ago, i.e. when `apply` method from `TableFunction` has been called.